python/ray/serve/_private/controller.py

```diff
@@ -431,6 +440,13 @@ async def run_control_loop(self) -> None:
                 except Exception:
                     logger.exception("Exception updating proxy state.")
 
+            # When the controller is done recovering, drop invalid handle metrics
+            # that may be stale for autoscaling
+            if not any_recovering:
+                self.deployment_state_manager.drop_invalid_handle_metrics(
+                    self.proxy_state_manager.get_alive_proxy_actor_ids()
+                )
```

Comment:
rather, I think we want to _avoid_ calling this prior to recovery finishing

Reply:
I believe when `any_recovering` is null, it is equivalent to recovery finishing.
And the call is made here instead of inside deployment state manager `update()` because it spans proxy and deployment info.